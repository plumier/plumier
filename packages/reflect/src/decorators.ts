import "reflect-metadata"

import { createClass, metadata } from "./helpers"
import { setMetadata } from "./metadata"
import {
    Class,
    CustomPropertyDecorator,
    DecoratorId,
    DecoratorOption,
    DecoratorOptionId,
    DecoratorTargetType,
    GenericTemplateDecorator,
    GenericTypeDecorator,
    NativeParameterDecorator,
    NoopDecorator,
    ParameterPropertiesDecorator,
    PrivateDecorator,
    TypeDecorator,
    TypeOverride,
} from "./types"


export function decorateParameter(callback: ((target: Class, name: string, index: number) => any), option?: DecoratorOption): ParameterDecorator
export function decorateParameter(data: any, option?: DecoratorOption): ParameterDecorator
export function decorateParameter(data: any, option?: DecoratorOption): ParameterDecorator {
    return (target: any, name: string | symbol, index: number) => {
        const isCtorParam = metadata.isConstructor(target)
        const targetClass = isCtorParam ? target : target.constructor
        const methodName = isCtorParam ? "constructor" : name
        const meta = typeof data === "function" ? data(targetClass, methodName, index) : data
        setMetadata({...meta, [DecoratorOptionId]: { ...meta[DecoratorOptionId], ...option }}, targetClass, methodName, index)
    }
}

export function decorateMethod(callback: ((target: Class, name: string) => any), option?: DecoratorOption): MethodDecorator
export function decorateMethod(data: any, option?: DecoratorOption): MethodDecorator
export function decorateMethod(data: any, option?: DecoratorOption) {
    return (target: any, name: string | symbol) => {
        const targetClass = target.constructor
        const meta = typeof data === "function" ? data(targetClass, name) : data
        setMetadata({...meta, [DecoratorOptionId]: { ...meta[DecoratorOptionId], ...option }}, targetClass, name)
    }
}

export function decorateProperty(callback: ((target: Class, name: string, index?: any) => any), option?: DecoratorOption): CustomPropertyDecorator
export function decorateProperty(data: any, option?: DecoratorOption): CustomPropertyDecorator
export function decorateProperty(data: any, option?: DecoratorOption) {
    return (target: any, name: string | symbol, index?: any) => {
        if (typeof index === "number")
            decorateParameter(data, option)(target, name, index)
        else
            decorateMethod(data, option)(target, name, index)
    }
}

export function decorateClass(callback: ((target: Class) => any), option?: DecoratorOption): ClassDecorator
export function decorateClass(data: any, option?: DecoratorOption): ClassDecorator
export function decorateClass(data: any, option?: DecoratorOption) {
    return (target: any) => {
        const meta = typeof data === "function" ? data(target) : data
        setMetadata({...meta, [DecoratorOptionId]: { ...meta[DecoratorOptionId], ...option }}, target)
    }
}

export function decorate(data: any | ((...args: any[]) => any), targetTypes: DecoratorTargetType[] = [], option?: Partial<DecoratorOption>) {
    const throwIfNotOfType = (target: DecoratorTargetType) => {
        if (targetTypes.length > 0 && !targetTypes.some(x => x === target))
            throw new Error(`Reflect Error: Decorator of type ${targetTypes.join(", ")} applied into ${target}`)
    }
    return (...args: any[]) => {
        //class decorator
        if (args.length === 1) {
            throwIfNotOfType("Class")
            return decorateClass(data, option)(args[0])
        }
        //parameter decorator
        if (args.length === 3 && typeof args[2] === "number") {
            throwIfNotOfType("Parameter")
            return decorateParameter(data, option)(args[0], args[1], args[2])
        }
        //property
        if (args[2] === undefined || args[2].get || args[2].set) {
            throwIfNotOfType("Property")
            return decorateProperty(data, option)(args[0], args[1], args[2])
        }
        throwIfNotOfType("Method")
        return decorateMethod(data, option)(args[0], args[1], args[2])
    }
}

export function mergeDecorator(...fn: (ClassDecorator | PropertyDecorator | NativeParameterDecorator | MethodDecorator)[]) {
    return (...args: any[]) => {
        fn.forEach(x => (x as Function)(...args))
    }
}

const symIgnore = Symbol("ignore")
const symOverride = Symbol("override")
const symParamProp = Symbol("paramProp")
const symNoop = Symbol("noop")

export function ignore() {
    return decorate(<PrivateDecorator>{ [DecoratorId]: symIgnore, kind: "Ignore" }, ["Parameter", "Method", "Property"], { allowMultiple: false })
}

export function type(type: TypeOverride | ((x: any) => TypeOverride), ...genericParams: (string | string[])[]) {
    return decorate((target: any) => <TypeDecorator>{ [DecoratorId]: symOverride, kind: "Override", type, genericParams: genericParams, target }, ["Parameter", "Method", "Property"], { inherit: true, allowMultiple: false })
}

export function noop() {
    // type is not inheritable because derived class can define their own type override
    return decorate((target: any) => <NoopDecorator>{ [DecoratorId]: symNoop, kind: "Noop", target }, undefined, { inherit: false, allowMultiple: false })
}

export function parameterProperties() {
    return decorateClass(<ParameterPropertiesDecorator>{ [DecoratorId]: symParamProp, type: "ParameterProperties" }, { allowMultiple: false })
}

export namespace generic {
    const symGenericType = Symbol("genericType")
    const symGenericTemplate = Symbol("genericTemplate")
    export function template(...templates: string[]) {
        return decorateClass(target => <GenericTemplateDecorator>{ [DecoratorId]: symGenericTemplate, kind: "GenericTemplate", templates, target }, { inherit: false, allowMultiple: false })
    }
    export function type(...types: TypeOverride[]) {
        return decorateClass(target => <GenericTypeDecorator>{ [DecoratorId]: symGenericType, kind: "GenericType", types, target }, { inherit: false, allowMultiple: false })
    }
    /**
     * Create generic class dynamically
     * @param parent Super class that the class inherited from
     * @param params List of generic type parameters
     */
    export function create<T extends Class>(parent: T | { parent: T, name: string }, ...params: TypeOverride[]) {
        const opt = (typeof parent === "object") ? parent : { parent: parent, name: "DynamicType" }
        const Type = createClass({ ...opt, genericParams: params })
        return Type
    }
}
