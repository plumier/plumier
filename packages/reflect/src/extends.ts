import { ClassReflection, MethodReflection, ParameterReflection, PropertyReflection } from "./types"

type MemberReflection = PropertyReflection | MethodReflection

function mergeMembers<T extends MemberReflection>(children: T[], parents: T[]): T[] {
    const result = [...children]
    for (const parent of parents) {
        if (!!result.find(x => x.name === parent.name)) continue
        result.push(parent)
    }
    return result
}

function extendsMetadata(child: ClassReflection, parent: ClassReflection): ClassReflection {
    return {
        ...child,
        methods: mergeMembers(child.methods, parent.methods),
        properties: mergeMembers(child.properties, parent.properties)
    }
}

export { extendsMetadata }