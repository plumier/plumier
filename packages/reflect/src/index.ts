import "reflect-metadata"

export * from "./types"
export * from "./decorators"
export * from "./helpers"
import { reflect } from "./reflect"
export { reflect }
export default reflect
