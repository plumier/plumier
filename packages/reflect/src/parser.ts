import { Node, parse } from "acorn"
import { getAllMetadata } from "./metadata"

import {
    Class,
    ClassReflection,
    ConstructorReflection,
    DecoratorOption,
    DecoratorOptionId,
    DECORATOR_KEY,
    FunctionReflection,
    MethodReflection,
    NativeDecorator,
    ParameterPropertyReflection,
    ParameterReflection,
    PropertyReflection,
} from "./types"

// --------------------------------------------------------------------- //
// ------------------------------ HELPERS ------------------------------ //
// --------------------------------------------------------------------- //

function getNode(node: Node, criteria: (x: any) => boolean): Node | undefined {
    if (criteria(node)) return node
    if (!(node as any).body) return
    if (Array.isArray((node as any).body)) {
        for (const child of (node as any).body) {
            const result = getNode(child, criteria)
            if (result) return result
        }
    }
    return getNode((node as any).body, criteria)
}

function getNamesFromAst(nodes: any[]) {
    const getName = (node: any): undefined | string | { [key: string]: string[] } => {
        if (node.type === "Identifier") return node.name
        if (node.type === "AssignmentPattern") return node.left.name
        if (node.type === "RestElement") return node.argument.name
        if (node.type === "Property") {
            if (node.value.type === "Identifier") return node.value.name
            else {
                const result: { [key: string]: any } = {}
                result[node.key.name] = getName(node.value)
                return result
            }
        }
        //if (node.type === "ObjectPattern") {
        return node.properties.map((x: any) => getName(x))
        //}
    }
    return nodes.map(x => getName(x)).filter((x): x is string | { [key: string]: string[] } => !!x)
}

function getCode(fn: Class | Function) {
    const code = fn.toString()
    if (code.search(/^class(\s*)extends\s*option.parent\s*{\s*}/gm) > -1) {
        return "class DynamicClass extends parent {}"
    }
    else
        return code.replace("[native code]", "")
}

function getMethodParameters(fn: Class, method: string) {
    const body = getCode(fn)
    const ast = parse(body, { ecmaVersion: 2020 })
    const ctor = getNode(ast, x => x.type === "MethodDefinition" && x.kind === "method" && x.key.name === method)
    return getNamesFromAst(ctor ? (ctor as any).value.params : [])
}

function getConstructorParameters(fn: Class) {
    const body = getCode(fn)
    const ast = parse(body, { ecmaVersion: 2020 })
    const ctor = getNode(ast, x => x.type === "MethodDefinition" && x.kind === "constructor")
    return getNamesFromAst(ctor ? (ctor as any).value.params : [])
}

function getFunctionParameters(fn: Function) {
    try {
        const body = getCode(fn)
        const ast = parse(body, { ecmaVersion: 2020 })
        return getNamesFromAst((ast as any).body[0].params)
    }
    catch {
        return []
    }
}

function getClassMembers(fun: Function) {
    const isGetter = (name: string) => Object.getOwnPropertyDescriptor(fun.prototype, name)!.get
    const isSetter = (name: string) => Object.getOwnPropertyDescriptor(fun.prototype, name)!.set
    const isFunction = (name: string) => typeof fun.prototype[name] === "function";
    const members = Object.getOwnPropertyNames(fun.prototype)
        .filter(name => isGetter(name) || isSetter(name) || isFunction(name))
    const properties = (getAllMetadata(fun as Class) || [])
        .filter(x => !!x.memberName && !x.parIndex)
        .filter(x => {
            const opt:DecoratorOption = x.data[DecoratorOptionId]
            return opt.applyTo!.length === 0
        })
        .map(x => x.memberName as string)
    const names = members.concat(properties)
        .filter(name => name !== "constructor" && !~name.indexOf("__"))
    return [...new Set(names)]
}

function printDestruct(params: any[]) {
    const result: string[] = []
    for (const key in params) {
        const par = params[key];
        if (typeof par === "string")
            result.push(par)
        else {
            const key = Object.keys(par)[0]
            result.push(`${key}: ${printDestruct(par[key])}`)
        }
    }
    return `{ ${result.join(", ")} }`
}

// --------------------------------------------------------------------- //
// ------------------------------- PARSER ------------------------------ //
// --------------------------------------------------------------------- //

function parseParameter(name: string | { [key: string]: string[] }, index: number): ParameterReflection {
    let parName
    let fields: { [key: string]: any[] } = {}
    if (typeof name === "object") {
        parName = printDestruct(name as any)
        fields = name
    }
    else {
        parName = name
    }
    return { kind: "Parameter", name: parName, decorators: [], fields, index, type: undefined }
}

function parseFunction(fn: Function): FunctionReflection {
    const parameters = getFunctionParameters(fn).map((x, i) => ({ ...parseParameter(x, i), type: undefined, typeClassification: undefined }))
    return { kind: "Function", name: fn.name, parameters, returnType: undefined }
}

function parseMethods(owner: Class): MethodReflection[] {
    const result: MethodReflection[] = []
    const members = getClassMembers(owner)
    for (const name of members) {
        const des = Reflect.getOwnPropertyDescriptor(owner.prototype, name)
        if (des && typeof des.value === "function" && !des.get && !des.set) {
            const parameters = getMethodParameters(owner, name).map((x, i) => parseParameter(x, i))
            result.push({ kind: "Method", name, parameters, decorators: [], returnType: undefined })
        }
    }
    return result
}

function parseProperties(owner: Class): PropertyReflection[] {
    const result: PropertyReflection[] = []
    const members = getClassMembers(owner)
    for (const name of members) {
        const des = Reflect.getOwnPropertyDescriptor(owner.prototype, name)
        if (!des || des.get || des.set) {
            result.push({ kind: "Property", name, decorators: [], get: des?.get, set: des?.set })
        }
    }
    // include constructor parameter
    const params = getConstructorParameters(owner)
    for (const [i, name] of params.entries()) {
        const { fields, ...par } = parseParameter(name, i)
        result.push(<ParameterPropertyReflection>{ ...par, kind: "Property", isParameter: true, get: undefined, set: undefined })
    }
    return result
}

function parseConstructor(fn: Class): ConstructorReflection {
    const params = getConstructorParameters(fn)
    return {
        kind: "Constructor",
        name: "constructor",
        parameters: params.map((x, i) => parseParameter(x, i)),
    }
}

function parseClass(fn: Class): ClassReflection {
    const proto = Object.getPrototypeOf(fn)
    return {
        kind: "Class", name: fn.name, type: fn, decorators: [],
        methods: parseMethods(fn),
        properties: parseProperties(fn),
        ctor: parseConstructor(fn),
        super: proto.prototype ? proto : Object,
    }
}

export { parseClass, parseFunction, getMethodParameters, getConstructorParameters, getFunctionParameters, getClassMembers }