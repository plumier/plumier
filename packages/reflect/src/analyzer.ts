import { generic } from "./decorators"
import { createClass, reflection } from "./helpers"
import { getMetadata, getMetadataForApplyTo, getOwnMetadata, mergeMetadata } from "./storage"
import {
    Class,
    DecoratorOption,
    DecoratorOptionId,
    DESIGN_PARAMETER_TYPE,
    DESIGN_RETURN_TYPE,
    DESIGN_TYPE,
    MethodReflection,
    ParameterPropertiesDecorator,
    ParameterPropertyReflection,
    ParameterReflection,
    PrivateDecorator,
    PropertyReflection,
    TypeDecorator,
} from "./types"
import { TypedReflection, WalkMemberContext } from "./walker"

// --------------------------------------------------------------------- //
// -------------------------- ADD DESIGN TYPE -------------------------- //
// --------------------------------------------------------------------- //

export function addsDesignTypes(meta: TypedReflection, ctx: WalkMemberContext): TypedReflection {
    const getType = (type: any, i: number) => type[i] === Array ? [Object] : type[i]
    if (meta.kind === "Method") {
        const returnType: any = Reflect.getOwnMetadata(DESIGN_RETURN_TYPE, ctx.target.prototype, meta.name)
        return { ...meta, returnType }
    }
    else if (reflection.isParameterProperties(meta)) {
        const parTypes: any[] = Reflect.getOwnMetadata(DESIGN_PARAMETER_TYPE, ctx.target) || []
        return { ...meta, type: getType(parTypes, meta.index) }
    }
    else if (meta.kind === "Property") {
        const type: any = Reflect.getOwnMetadata(DESIGN_TYPE, ctx.target.prototype, meta.name)
        return { ...meta, type }
    }
    else if (meta.kind === "Parameter" && ctx.parent.kind === "Constructor") {
        const parTypes: any[] = Reflect.getOwnMetadata(DESIGN_PARAMETER_TYPE, ctx.target) || []
        return { ...meta, type: getType(parTypes, meta.index) }
    }
    else if (meta.kind === "Parameter" && ctx.parent.kind === "Method") {
        const parTypes: any[] = Reflect.getOwnMetadata(DESIGN_PARAMETER_TYPE, ctx.target.prototype, ctx.parent.name) || []
        return { ...meta, type: getType(parTypes, meta.index) }
    }
    else
        return meta
}

// --------------------------------------------------------------------- //
// --------------------------- ADD DECORATORS -------------------------- //
// --------------------------------------------------------------------- //

export function addsDecorators(meta: TypedReflection, ctx: WalkMemberContext): TypedReflection {
    if (meta.kind === "Parameter") {
        // constructor metadata are not inheritable
        const decorators = ctx.parent.name === "constructor"
            ? getOwnMetadata(ctx.target, ctx.parent.name, meta.index)
            : getMetadata(ctx.target, ctx.parent.name, meta.index)
        return { ...meta, decorators: meta.decorators.concat(decorators) }
    }
    if (reflection.isParameterProperties(meta)) {
        // get copy own metadata of constructor 
        const decorators = getOwnMetadata(ctx.target, "constructor", meta.index)
            // and also a copy of metadata of the property (using applyTo)
            .concat(...getMetadata(ctx.target, meta.name))
        return { ...meta, decorators: meta.decorators.concat(decorators) }
    }
    if (meta.kind === "Method" || meta.kind === "Property") {
        const decorators = getMetadata(ctx.target, meta.name)
        return { ...meta, decorators: meta.decorators.concat(decorators) }
    }
    if (meta.kind === "Class") {
        const decorators = getMetadata(meta.type)
        return { ...meta, decorators: meta.decorators.concat(decorators) }
    }
    return meta
}

// --------------------------------------------------------------------- //
// ----------------------- ADD APPLY-TO DECORATOR ---------------------- //
// --------------------------------------------------------------------- //

export function addsApplyToDecorator(meta: TypedReflection, ctx: WalkMemberContext) {
    if (reflection.isParameterProperties(meta)) {
        // get copy own metadata of constructor 
        const decorators = getMetadataForApplyTo(ctx.target, "constructor", meta.index)
            // and also a copy of metadata of the property (using applyTo)
            .concat(...getMetadataForApplyTo(ctx.target, meta.name))
        return { ...meta, decorators: meta.decorators.concat(decorators) }
    }
    if (meta.kind === "Method" || meta.kind === "Property") {
        const decorators = getMetadataForApplyTo(ctx.target, meta.name)
        return { ...meta, decorators: mergeMetadata(decorators, meta.decorators, false) }
    }
    if (meta.kind === "Class") {
        const rawDecorators = getMetadataForApplyTo(meta.type)
        const removedDecorators = rawDecorators.filter(x => {
            const option: Required<DecoratorOption> = x[DecoratorOptionId]
            return option.removeApplied
        })
        const decorators = rawDecorators.filter(x => {
            const option: Required<DecoratorOption> = x[DecoratorOptionId]
            return !option.removeApplied
        })
        const result = { ...meta, decorators: meta.decorators.concat(decorators) }
        if (removedDecorators.length > 0)
            result.removedDecorators = removedDecorators
        return result
    }
    return meta
}

// --------------------------------------------------------------------- //
// ------------------ ADD DATA TYPE BY TYPE DECORATOR ------------------ //
// --------------------------------------------------------------------- //

// check if a generic type @type("T")
function isGeneric(decorator: TypeDecorator) {
    const [singleType] = reflection.getTypeFromDecorator(decorator)
    return typeof singleType === "string"
}

// check if a dynamic type @type({ name: String, age: Number })
function isDynamic(decorator: TypeDecorator) {
    const [singleType] = reflection.getTypeFromDecorator(decorator)
    return typeof singleType === "object"
}

// check if a type with generic parameter @type(Type, "T", "U")
function isTypeWithGenericParameter(decorator: TypeDecorator) {
    const [singleType] = reflection.getTypeFromDecorator(decorator)
    return typeof singleType === "function" && decorator.genericParams.length > 0
}

function setType(meta: MethodReflection | PropertyReflection | ParameterReflection | ParameterPropertyReflection, type: Class | Class[]) {
    if (meta.kind === "Method")
        return { returnType: type }
    return { type }
}

export function addsTypeByTypeDecorator(meta: TypedReflection, ctx: WalkMemberContext): TypedReflection {
    if (meta.kind === "Constructor" || meta.kind === "Class") return meta
    const decorator = meta.decorators.find((x: TypeDecorator): x is TypeDecorator => x.kind === "Override")
    if (!decorator) return meta
    if (isGeneric(decorator)) {
        const type = generic.getType(decorator, ctx.target)!
        return { ...meta, ...setType(meta, type) }
    }
    if (isDynamic(decorator)) {
        const [rawType, isArray] = reflection.getTypeFromDecorator(decorator)
        const type = createClass(rawType as {})
        return { ...meta, ...setType(meta, isArray ? [type] : type as any) }
    }
    if (isTypeWithGenericParameter(decorator)) {
        const genericParams: any[] = decorator.genericParams.map(x => generic.getType({ type: x, target: decorator.target }, ctx.target))
        const [parentType, isArray] = reflection.getTypeFromDecorator(decorator)
        const dynType = createClass({}, { extends: parentType as any, genericParams })
        const type = isArray ? [dynType] : dynType
        return { ...meta, ...setType(meta, type) }
    }
    const [type, isArray] = reflection.getTypeFromDecorator(decorator)
    return { ...meta, ...setType(meta, isArray ? [type] : type as any) }
}

// --------------------------------------------------------------------- //
// ----------------------- ADD TYPE CLASSIFICATION --------------------- //
// --------------------------------------------------------------------- //

export function addsTypeClassification(meta: TypedReflection, ctx: WalkMemberContext): TypedReflection | undefined {
    const get = (type: any): "Class" | "Array" | "Primitive" | undefined => {
        if (type === undefined) return undefined
        else if (Array.isArray(type)) return "Array"
        else if (reflection.isCustomClass(type)) return "Class"
        else return "Primitive"
    }
    if (meta.kind === "Method")
        return { ...meta, typeClassification: get(meta.returnType) }
    else if (meta.kind === "Property" || meta.kind === "Parameter")
        return { ...meta, typeClassification: get(meta.type) }
    else if (meta.kind === "Class")
        return { ...meta, typeClassification: "Class" }
    return meta
}

// --------------------------------------------------------------------- //
// ---------------------- ADD PARAMETER PROPERTIES --------------------- //
// --------------------------------------------------------------------- //

export function addsParameterProperties(meta: TypedReflection, ctx: WalkMemberContext): TypedReflection | undefined {
    if (reflection.isParameterProperties(meta) && ctx.parent.kind === "Class") {
        const isParamProp = ctx.parent.decorators.some((x: ParameterPropertiesDecorator) => x.type === "ParameterProperties")
        return !!isParamProp ? meta : undefined
    }
    return meta
}

// --------------------------------------------------------------------- //
// --------------------------- REMOVE IGNORE --------------------------- //
// --------------------------------------------------------------------- //

export function removeIgnored(meta: TypedReflection, ctx: WalkMemberContext): TypedReflection | undefined {
    if (meta.kind === "Property" || meta.kind === "Method") {
        const decorator = meta.decorators.find((x: PrivateDecorator): x is PrivateDecorator => x.kind === "Ignore")
        return !decorator ? meta : undefined
    }
    return meta
}
