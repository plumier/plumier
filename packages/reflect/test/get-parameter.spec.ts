import {decorate} from "../src/decorators"
import * as metadata from "../src/parser"

function globalFunction(a: any, b: any) {

}

describe("Constructor parameter", () => {
    it("Should get constructor parameter name", () => {
        class DummyClass {
            constructor(
                par1: string,
                par2: string,
                par3: string,
                par4: string
            ) {
                globalFunction(par1, par2)
            }
        }
        const result = metadata.getConstructorParameters(DummyClass)
        expect(result).toMatchSnapshot()
    })

    it("Should get constructor parameter name with comment", () => {
        class DummyClass {
            constructor(
                /* this is comment */
                par1: string,
                /* this is comment () */
                par2: string,
                par3: string,
                /* this is comment {} */
                par4: string
            ) {
                globalFunction(par1, par2)
            }
        }
        const result = metadata.getConstructorParameters(DummyClass)
        expect(result).toMatchSnapshot()
    })

    it("Should get constructor with default parameter", () => {
        class DummyClass {
            constructor(
                par1 = "Halo",
                par2 = 123,
                par3 = new Date(),
                par4 = false
            ) {
                globalFunction(par1, par2)
            }
        }
        const result = metadata.getConstructorParameters(DummyClass)
        expect(result).toMatchSnapshot()
    })

    it("Should reflect rest parameter", () => {
        class DummyClass {
            constructor(...pars: any[]) { }
        }
        const result = metadata.getConstructorParameters(DummyClass)
        expect(result).toMatchSnapshot()
    })

    it("Should return empty array when no constructor provided", () => {
        class DummyClass {
            @decorate({})
            myProp = 1;
            @decorate({})
            myFunction(
                @decorate({}) par: string
            ) { }
        }
        const result = metadata.getConstructorParameters(DummyClass)
        expect(result).toMatchSnapshot()
    })

})

describe("Method Parameters", () => {
    it("Should get method parameter name", () => {
        class DummyClass {
            myMethod(
                par1: string,
                par2: string,
                par3: string,
                par4: string
            ) {
                globalFunction(par1, par2)
            }
        }
        const result = metadata.getMethodParameters(DummyClass, "myMethod")
        expect(result).toMatchSnapshot()
    })

    it("Should get method parameter name with comment", () => {
        class DummyClass {
            myMethod(
                /* this is comment */
                par1: string,
                /* this is comment () */
                par2: string,
                par3: string,
                /* this is comment {} */
                par4: string
            ) {
                globalFunction(par1, par2)
            }
        }
        const result = metadata.getMethodParameters(DummyClass, "myMethod")
        expect(result).toMatchSnapshot()
    })

    it("Should get method parameter with default parameter", () => {
        class DummyClass {
            myMethod(
                par1 = "Halo",
                par2 = 123,
                par3 = new Date(),
                par4 = false
            ) {
                globalFunction(par1, par2)
            }
        }
        const result = metadata.getMethodParameters(DummyClass, "myMethod")
        expect(result).toMatchSnapshot()
    })
})

describe("Function Parameters", () => {
    it("Should get function parameter name", () => {
        function myFunction(
            par1: string,
            par2: string,
            par3: string,
            par4: string
        ) {
            globalFunction(par1, par2)
        }
        const result = metadata.getFunctionParameters(myFunction)
        expect(result).toMatchSnapshot()
    })

    it("Should get function parameter name with comment", () => {
        function myFunction(
            /* this is comment */
            par1: string,
            /* this is comment () */
            par2: string,
            par3: string,
            /* this is comment {} */
            par4: string
        ) {
            globalFunction(par1, par2)
        }
        const result = metadata.getFunctionParameters(myFunction)
        expect(result).toMatchSnapshot()
    })

    it("Should get function parameter with default parameter", () => {
        function myFunction(
            par1 = "Halo",
            par2 = 123,
            par3 = new Date(),
            par4 = false
        ) {
            globalFunction(par1, par2)
        }
        const result = metadata.getFunctionParameters(myFunction)
        expect(result).toMatchSnapshot()
    })
})

describe("Durability", () => {
    it("getParameterNames should not error when provided proxy", () => {
        function MyFunction() { }
        MyFunction.toString = () => {
            return "[Function]"
        }
        const result = metadata.getFunctionParameters(MyFunction)
        expect(result).toMatchSnapshot()
    })

    it("getConstructorParameters should not error when provided proxy", () => {
        class MyFunction { 
            constructor(){}
        }
        MyFunction.toString = () => {
            return "[Function]"
        }
        const result = metadata.getConstructorParameters(MyFunction)
        expect(result).toMatchSnapshot()
    })

    it("getMethodParameters should not error when provided proxy", () => {
        class MyFunction { 
            constructor(){}
            myMethod(){}
        }
        MyFunction.toString = () => {
            return "[Function]"
        }
        const result = metadata.getMethodParameters(MyFunction, "myMethod")
        expect(result).toMatchSnapshot()
    })

    it("Should not error when provided function without parameter", () => {
        function myFun() { }
        const result = metadata.getFunctionParameters(myFun)
        expect(result).toMatchSnapshot()
    })

    it("Should not error when provided method without parameter", () => {
        class MyClass {
            myMethod() { }
        }
        const result = metadata.getMethodParameters(MyClass, "myMethod")
        expect(result).toMatchSnapshot()
    })

    it("Should not error when provided constructor without parameter", () => {
        class MyClass {
            constructor() { }
        }
        const result = metadata.getConstructorParameters(MyClass)
        expect(result).toMatchSnapshot()
    })

    it("Should not error when provided default constructor", () => {
        class MyClass {
        }
        const result = metadata.getConstructorParameters(MyClass)
        expect(result).toMatchSnapshot()
    })

    it("Should not error when provided method first than constructor", () => {
        class MyClass {
            myMethod() { }
            constructor() { }
        }
        const result = metadata.getConstructorParameters(MyClass)
        expect(result).toMatchSnapshot()
    })
})

describe("Parameter Destructuring", () => {
    it("Should parse parameter destructuring", () => {
        interface MyModel {
            date: Date
            num: number
        }
        function myFun(par:string, { date, num }: MyModel) { }
        const result = metadata.getFunctionParameters(myFun)
        expect(result).toMatchSnapshot()
    })

    it("Should parse parameter destructuring with custom name", () => {
        interface MyModel {
            date: Date
            num: number
        }
        function myFun({ date: tanggal, num }: MyModel) { }
        const result = metadata.getFunctionParameters(myFun)
        expect(result).toMatchSnapshot()
    })

    it("Should parse parameter destructuring with nested property", () => {
        interface MyModel {
            date: Date
            num: number,
            inner: InnerModel
        }
        interface InnerModel {
            str: string
            dob: Date
        }
        function myFun(par:string, { date: tanggal, num, inner: { str, dob: dateOfBirth } }: MyModel) { }
        const result = metadata.getFunctionParameters(myFun)
        expect(result).toMatchSnapshot()
    })

    it("Should parse parameter destructuring on constructor", () => {
        interface MyModel {
            date: Date
            num: number
        }
        class MyClass {
            constructor({ date: tanggal, num }: MyModel) { }
        }
        const result = metadata.getConstructorParameters(MyClass)
        expect(result).toMatchSnapshot()
    })

    it("Should parse parameter destructuring on method", () => {
        interface MyModel {
            date: Date
            num: number
        }
        class MyClass {
            myMethod({ date: tanggal, num }: MyModel) { }
        }
        const result = metadata.getMethodParameters(MyClass, "myMethod")
        expect(result).toMatchSnapshot()
    })
})