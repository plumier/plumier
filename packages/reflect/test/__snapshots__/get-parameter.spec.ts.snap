// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Constructor parameter Should get constructor parameter name 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Constructor parameter Should get constructor parameter name with comment 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Constructor parameter Should get constructor with default parameter 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Constructor parameter Should reflect rest parameter 1`] = `
Array [
  "pars",
]
`;

exports[`Constructor parameter Should return empty array when no constructor provided 1`] = `Array []`;

exports[`Durability Should not error when provided constructor without parameter 1`] = `Array []`;

exports[`Durability Should not error when provided default constructor 1`] = `Array []`;

exports[`Durability Should not error when provided function without parameter 1`] = `Array []`;

exports[`Durability Should not error when provided method first than constructor 1`] = `Array []`;

exports[`Durability Should not error when provided method without parameter 1`] = `Array []`;

exports[`Durability getConstructorParameters should not error when provided proxy 1`] = `Array []`;

exports[`Durability getMethodParameters should not error when provided proxy 1`] = `Array []`;

exports[`Durability getParameterNames should not error when provided proxy 1`] = `Array []`;

exports[`Function Parameters Should get function parameter name 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Function Parameters Should get function parameter name with comment 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Function Parameters Should get function parameter with default parameter 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Method Parameters Should get method parameter name 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Method Parameters Should get method parameter name with comment 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Method Parameters Should get method parameter with default parameter 1`] = `
Array [
  "par1",
  "par2",
  "par3",
  "par4",
]
`;

exports[`Parameter Destructuring Should parse parameter destructuring 1`] = `
Array [
  "par",
  Array [
    "date",
    "num",
  ],
]
`;

exports[`Parameter Destructuring Should parse parameter destructuring on constructor 1`] = `
Array [
  Array [
    "tanggal",
    "num",
  ],
]
`;

exports[`Parameter Destructuring Should parse parameter destructuring on method 1`] = `
Array [
  Array [
    "tanggal",
    "num",
  ],
]
`;

exports[`Parameter Destructuring Should parse parameter destructuring with custom name 1`] = `
Array [
  Array [
    "tanggal",
    "num",
  ],
]
`;

exports[`Parameter Destructuring Should parse parameter destructuring with nested property 1`] = `
Array [
  "par",
  Array [
    "tanggal",
    "num",
    Object {
      "inner": Array [
        "str",
        "dateOfBirth",
      ],
    },
  ],
]
`;
